src/FlaUI.Core/AutomationElements/AutomationElementExtensions.cs

```diff
@@ -73,5 +74,32 @@ public static partial class AutomationElementExtensions
             }
             return self;
         }
+
+        /// <summary>
+        /// Clicks in chosen element, creates instance of target object and returns it.
+        /// </summary>
+        /// <typeparam name="T">Type of target object to create.</typeparam>
+        public static T Click<T>(this AutomationElement element, TimeSpan? timeout = null, TimeSpan? interval = null) where T : AutomationElement
+        {
+            Retry.WhileException(() =>
+                {
+                    element.Click();
+                    Wait.UntilInputIsProcessed();
+                },
+                timeout,
+                interval,
+                true);
+
+            var type = typeof(T);
+            var ctorAutomationElement = type.GetConstructor(new[] { typeof(AutomationElement) });
+
+            return Retry.WhileNull(
+                    () => (T)ctorAutomationElement?.Invoke(new object[] { element }),
```

Comment:
While using Page Object Pattern it makes a lot of sense. When Click() doesn't return void, but some specific object, you can act on it right away and chain calls. So for example you could write something like this:
`_unitUnderTest = LoginButton.Click<ProfilePage>().EditProfile.Click<ProfileEditPage>();`
and so on...

In my opinin, it helps structurize tests, keeps it a little more cleaner and readable (no need to create variables for each transistion between views).

Regarding usage of Retry.Whilenull() - it's just to be sure that target element is ready, just in case rendering would be too slow. 

And about null - parameter is restricted to AutomationElement
`public static T Click<T>(...) where T : AutomationElement`
so as long it is of type AutomationElement or is inheriting, we're safe ;)

Reply:
Thanks for elaborating, it's more clear now. Regarding Retry.WhileNull now I see it - AutomationElement ctor throws an exception when null is passed and ignoreException parameter of WhileNull is set to true so it really makes sense, cool :)

Reply:
Thanks for elaborating, it's more clear now. Regarding Retry.WhileNull I see it now - AutomationElement ctor throws an exception when null is passed and ignoreException parameter of WhileNull is set to true so it really makes sense, cool :)

Reply:
After giving it a second though there is one more thing I don't get - how can you create ProfilePage element from Button element? Is there some magic in ProfilePage ctor?

Reply:
unfortunately, no magic ;)
you just need to pass parent element/window to ProfilePage ctor to be in the right scope, so the page's ID (or whatever marker you're looking for) can be found.

Reply:
@michal-ra what do you mean by "pass"? You are passing `automationElement` parameter to element you're creating so in your example Button's automationElement will be passed to ProfilePage, won't it?

Reply:
@michal-ra ^

Reply:
@michal-ra ^

Reply:
Same. I don't see how that is supposed to make anything easier. respectively more transparent. Usually in UI testing you try to keep every step at it's own so it is easier to debug and log.